### PR TITLE
config: make docker health check configurable

### PR DESCRIPTION
**Pull request by adnxn**
### Summary

carried over from https://github.com/aws/amazon-ecs-agent/pull/1522

### Testing

- [ ] Builds on Linux (`make release`)
- [ ] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [ ] Unit tests on Linux (`make test`) pass
- [ ] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [ ] Integration tests on Linux (`make run-integ-tests`) pass
- [ ] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [ ] Functional tests on Linux (`make run-functional-tests`) pass
- [ ] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/tcs/handler/types.go [hunk 1]
```diff
@@ -50,3 +51,10 @@ func (params *TelemetrySessionParams) time() ttime.Time {
 	})
 	return params._time
 }
+
+func (params *TelemetrySessionParams) isMetricsDisabled() (bool, error) {
```
Comment (yumex93) [inline agent/tcs/handler/types.go]: The function name makes me a little bit confusing. It is called isMetricsDisabled, but when DisableMetrics flag is true, the function will not return true. Do you want to say MetricsSessionDisabled or something? Maybe change the function name and add some description as comments here.

File: agent/tcs/handler/types_test.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+			result:      false,
+			err:         nil,
+			description: "Only telemetry metrics was disable should return false",
+		},
+		{
+			param:       &TelemetrySessionParams{Cfg: &config.Config{DisableMetrics: true, DisableDockerHealthCheck: true}},
+			result:      true,
+			err:         nil,
+			description: "both telemetry and health metrics were disable should return true",
+		},
```
Comment (yumex93) [inline agent/tcs/handler/types_test.go]: Is it in golang, for **&&** operator, if the first one is false, it will not proceed to check the value of second parameter? Since seems you miss a case, when **DisableMetrics** is false and **DisableDockerHealthCheck** is true.

File: agent/app/agent_capability_test.go [hunk 3]
```diff
@@ -587,6 +587,43 @@ func TestCapabilitiesContainerHealth(t *testing.T) {
...
+	client.EXPECT().SupportedVersions().Return([]dockerclient.DockerVersion{
+		dockerclient.Version_1_24,
+	})
+	client.EXPECT().KnownVersions().Return(nil)
+	mockMobyPlugins.EXPECT().Scan().AnyTimes().Return([]string{}, nil)
+	client.EXPECT().ListPluginsWithFilters(gomock.Any(), gomock.Any(), gomock.Any(),
+		gomock.Any()).AnyTimes().Return([]string{}, nil)
+
+	ctx, cancel := context.WithCancel(context.TODO())
+	// Cancel the context to cancel async routines
```
Comment (fenxiong) [inline agent/app/agent_capability_test.go]: why is this needed?

File: agent/app/agent_capability_test.go [hunk 4]
```diff
@@ -587,6 +587,43 @@ func TestCapabilitiesContainerHealth(t *testing.T) {
...
+		ctx:          ctx,
+		cfg:          &config.Config{DisableDockerHealthCheck: true},
+		dockerClient: client,
+		mobyPlugins:  mockMobyPlugins,
+	}
+
+	capabilities, err := agent.capabilities()
+	require.NoError(t, err)
+
+	capMap := make(map[string]bool)
```
Comment (fenxiong) [inline agent/app/agent_capability_test.go]: i think we can use [assert.NotContains](https://godoc.org/github.com/pgpst/pgpst/internal/github.com/stretchr/testify/assert#NotContains) instead of these

File: agent/tcs/handler/handler.go [hunk 5]
```diff
@@ -47,7 +47,17 @@ const (
...
 // and invokes StartSession.
 func StartMetricsSession(params TelemetrySessionParams) {
-	err := params.StatsEngine.MustInit(params.Ctx, params.TaskEngine, params.Cfg.Cluster,
+	ok, err := params.isMetricsDisabled()
+	if err != nil {
+		seelog.Warnf("Error starting metrics session: %v", err)
+		return
+	}
+	if ok {
+		seelog.Warnf("Metrics were disabled, not start the telemetry session")
```
Comment (fenxiong) [inline agent/tcs/handler/handler.go]: nit: start -> starting

File: agent/tcs/handler/handler_test.go [hunk 6]
```diff
@@ -73,6 +73,19 @@ func (*mockStatsEngine) GetTaskHealthMetrics() (*ecstcs.HealthMetadata, []*ecstc
 	return nil, nil, nil
 }
 
+// TestDisableMetrics tests the StartMetricsSession will return immediately if
```
Comment (fenxiong) [inline agent/tcs/handler/handler_test.go]: just wondering what is the result of breaking the logic we test here? it seems that in that case we will just not return immediately, but i'm not sure whether the test will fail?

File: agent/tcs/handler/types_test.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.description, func(t *testing.T) {
+			ok, err := tc.param.isMetricsDisabled()
+			if tc.err != nil {
+				assert.Error(t, err)
+			}
```
Comment (fenxiong) [inline agent/tcs/handler/types_test.go]: else we should assert no error?

File: agent/tcs/handler/types.go [hunk 8]
```diff
@@ -50,3 +51,10 @@ func (params *TelemetrySessionParams) time() ttime.Time {
...
+	if params.Cfg != nil {
+		return params.Cfg.DisableMetrics && params.Cfg.DisableDockerHealthCheck, nil
```
Comment (haikuoliu) [inline agent/tcs/handler/types.go]: Just for understanding, is `DisableMetrics` something that we have introduced previously but never used? Would it be better to add a description in the PR summary for it?

**Comment by adnxn** [comment]
thanks for taking a pass at this, will make changes and need to update the test for this actually as well.

File: agent/tcs/handler/types.go (hunk 1 above)
Comment (adnxn) [reply agent/tcs/handler/types.go]: fair. changed to `isContainerHealthMetricsDisabled` since checking against both `DisabledMetrics && DisableDockerHealthCheck`

File: agent/app/agent_capability_test.go (hunk 3 above)
Comment (adnxn) [reply agent/app/agent_capability_test.go]: without it we end up with flakey tests that have run away go routines. maybe you've seen some errors similar to "goroutine ended after test" kind of thing. please correct me if i'm wrong though.

File: agent/tcs/handler/handler_test.go (hunk 6 above)
Comment (adnxn) [reply agent/tcs/handler/handler_test.go]: hrm yea this is a weird one, this looks like it needs a `StartMetricsSession` refactor to be able to have more sensible testing. we can track this as tech debt.

File: agent/tcs/handler/types.go (hunk 8 above)
Comment (adnxn) [reply agent/tcs/handler/types.go]: `DisableMetrics` maps to `ECS_DISABLE_METRICS` and is already covered in the README. this PR is to add a toggle for the health checks, which relies on the underlying tcs/metrics subsystem.

File: agent/tcs/handler/handler_test.go (hunk 6 above)
Comment (fenxiong) [reply agent/tcs/handler/handler_test.go]: nit: add a TODO for this

**Comment by adnxn** [comment]
windows ftest failing:

```
=== RUN   TestTwoTasksSharedLocalVolume
go : panic: test timed out after 30m0s
At C:\Users\Administrator\AppData\Local\Temp\amazon-ecs-agent\go\src\github.com\aws\amazon-ecs-agent\scripts\run-functi
onal-tests.ps1:26 char:3
+   go test -tags functional -timeout=30m -v ../agent/functional_tests/ ...
+   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (panic: test timed out after 30m0s:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
```
